### PR TITLE
Integrate Custom Random Matrix Generation and Merge Ubuntu and Windows Workflows

### PR DESCRIPTION
**Pull request by BereKanters**
This PR addresses issue #1460 by integrating a custom random matrix generation script and merging the Ubuntu and Windows workflows into a single CI workflow. This change aims to improve test coverage and CI resource efficiency.

- Combined the Ubuntu and Windows workflows into a single ci.yml file.
- Moved common steps, such as caching the Android jar and compiling, into a separate setup job to improve CI efficiency.
- Updated the README file to reflect the new combined CI workflow badge.
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/main/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -12,50 +11,17 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v4
```
Comment (robstoll) [inline .github/workflows/ci.yml:12]:
```suggestion
        uses: actions/checkout@v4
        with:
          fetch-depth: 1
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -12,50 +11,17 @@ jobs:
...
-  setup:
-    runs-on: ubuntu-latest
-    steps:
-      - name: Checkout
-        uses: actions/checkout@v4
-
-      - name: Cache Android jar
-        id: android-cache
-        uses: actions/cache@v4
+      - name: Set up Node.js
```
Comment (robstoll) [inline .github/workflows/ci.yml:14]: no need to setup node, node is installed already for each runner

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -83,23 +49,51 @@ jobs:
...
+            sha256sum android.jar
+            exit 1;
+          fi
+
+      - name: Check Atrium's -jvm.jar can be dexed
+        run: ATRIUM_ANDROID_JAR="$PWD/android-jar-cache/android.jar" ./gradlew checkDexer
+
   forwardCompatibility:
-    needs: setup
+    needs: dexer
```
Comment (robstoll) [inline .github/workflows/ci.yml:88]: shouldn't be based on dexer, can be run in parallel to it

**Review by robstoll — CHANGES_REQUESTED** [review]

File: feature-random-matrix-ci.patch
Comment (robstoll) [inline feature-random-matrix-ci.patch:1]: please remove this file again

File: .github/workflows/ci.yml
Comment (robstoll) [inline .github/workflows/ci.yml:1]: please rename to quality-assurance.yml (and adopt the badges in the readme, I'll leave an example)

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    strategy:
+      matrix: ${{ fromJson(needs.generate-matrix.outputs.matrix) }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up JDK ${{ matrix.java_version }}
+        uses: actions/setup-java@v4
+        with:
+          distribution: 'adopt'
```
Comment (robstoll) [inline .github/workflows/ci.yml:35]:
```suggestion
          distribution: ${{ matrix.java_distribution }}
```

**Review by robstoll — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -3,8 +3,7 @@
 [![Download](https://img.shields.io/badge/Download-1.2.0-%23007ec6)](https://central.sonatype.com/artifact/ch.tutteli.atrium/atrium-fluent/1.2.0)
 [![EUPL](https://img.shields.io/badge/%E2%9A%96-EUPL%201.2-%230b45a6)](https://joinup.ec.europa.eu/collection/eupl/eupl-text-11-12 "License")
 [![atrium @ kotlinlang.slack.com](https://img.shields.io/static/v1?label=kotlinlang&message=atrium&color=blue&logo=slack)](https://kotlinlang.slack.com/messages/atrium "See invitation link under section FAQ")
-[![Build Status Ubuntu](https://github.com/robstoll/atrium/workflows/Ubuntu/badge.svg?event=push&branch=main)](https://github.com/robstoll/atrium/actions?query=workflow%3AUbuntu+branch%3Amain)
-[![Build Status Windows](https://github.com/robstoll/atrium/workflows/Windows/badge.svg?event=push&branch=main)](https://github.com/robstoll/atrium/actions?query=workflow%3AWindows+branch%3Amain)
+[![Build Status](https://github.com/robstoll/atrium/actions/workflows/ci.yml/badge.svg?event=push&branch=main)](https://github.com/robstoll/atrium/actions?query=workflow%3Aci.yml+branch%3Amain)
```
Comment (robstoll) [inline README.md:6]:
```suggestion
[![Quality Assurance](https://github.com/robstoll/atrium/actions/workflows/quality-assurance.yml/badge.svg?event=push&branch=main)](https://github.com/robstoll/atrium/actions/workflows/quality-assurance.yml?query=branch%3Amain)
```

File: .github/workflows/ci.yml
Comment (robstoll) [inline .github/workflows/ci.yml:1]: Don't forget to remove the build-ubuntu.yml and build-windows.yml in the end

File: .github/workflows/quality-assurance.yml [hunk 6]
```diff
@@ -85,22 +86,22 @@ jobs:
         run: ATRIUM_ANDROID_JAR="$PWD/android-jar-cache/android.jar" ./gradlew checkDexer
 
   forwardCompatibility:
-    needs: dexer
+    needs: []  #Not sure if it should be left empty or if it should contain generate-matrix
```
Comment (robstoll) [inline .github/workflows/quality-assurance.yml:89]: if you would want to use different OS etc. then you would need it. IMO good enough if we use jdk17 and adopt only for now

File: .github/workflows/quality-assurance.yml [hunk 7]
```diff
@@ -85,22 +86,22 @@ jobs:
...
-        kotlin_version: ['1.5', '1.6', '1.7', '1.8', '1.9', '2.0']
+        kotlin_version: [ '1.5', '1.6', '1.7', '1.8', '1.9', '2.0' ]
     steps:
       - uses: actions/checkout@v4
       - name: Set up JDK 17
         uses: actions/setup-java@v4
         with:
-          distribution: 'adopt'
-          java-version: 17
+          distribution: ${{ matrix.java_distribution }}
```
Comment (robstoll) [inline .github/workflows/quality-assurance.yml:103]: keep adopt and 17 we define the matrix differently in this job (i.e. we only define `kotlin_version`, see further above)

File: .github/workflows/quality-assurance.yml [hunk 8]
```diff
@@ -1,56 +1,79 @@
...
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix: ${{ fromJson(needs.generate-matrix.outputs.matrix) }}
     steps:
-      - uses: actions/checkout@v4
-      - name: "Set up JDK ${{ matrix.java_version }}"
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
```
Comment (robstoll) [inline .github/workflows/quality-assurance.yml:29]: I think codecov needs the history to work correctly, remove here again (but keep the others)

File: .github/workflows/quality-assurance.yml [hunk 9]
```diff
@@ -1,56 +1,79 @@
...
+        run: ./gradlew build
+        working-directory: misc/tools/atrium-samples-test
+
+  dexer:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
```
Comment (robstoll) [inline .github/workflows/quality-assurance.yml:61]: good idea, we also don't need the history for the dexer job :+1:

File: .github/workflows/quality-assurance.yml [hunk 10]
```diff
@@ -86,7 +84,7 @@ jobs:
         run: ATRIUM_ANDROID_JAR="$PWD/android-jar-cache/android.jar" ./gradlew checkDexer
 
   forwardCompatibility:
-    needs: []  #Not sure if it should be left empty or if it should contain generate-matrix
+    needs: []
```
Comment (robstoll) [inline .github/workflows/quality-assurance.yml:87]: You could also remove the line entirely

**Review by robstoll — COMMENTED** [review]
You have merge conflicts which means you need to rebase on main of Atrium, i.e. not on your local main -- add atrium as `git remote add upstream https://github.com/robstoll/atrium` and then `git rebase upstream/main` and resolve the conflicts. Let me know in case you need help with it

**Comment by BereKanters** [comment]
I have no clue what just happend, I did git rebase upstream/main (I already had the upstream set) afterwords asked me to resolve conflicts where I removed the build_windows file by using: 'git rm .github/workflows/build-windows.yml ' then added `git add .github/workflows/quality-assurance.yml` did `git rebase --continue ` and then pushed but it seems it pushed all of that with it for some reason.

**Comment by BereKanters** [comment]
Alright I went back to an older commit [28ff511] and it should now be fixed.

File: .github/workflows/matrix.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+// Add axes for the matrix
+matrix.addAxis({
+    name: 'os',
+    title: x => x.replace('-latest', ''),
+    values: [
+        'ubuntu-latest',
+        'windows-latest'
+    ]
+});
+matrix.addAxis({
```
Comment (robstoll) [inline .github/workflows/matrix.js:13]: You also need to add an axis for distribution. I suggest you use:
'corretto',
		'liberica',
		'microsoft',
		'temurin',
		'zulu',

**Review by robstoll — DISMISSED** [review]

**Comment by robstoll** [comment]
@BereKanters thanks for your second contribution. I missed that you changed the target branch just before I merged. We'll figure out if everything was as intended at a screen bigger than my mobile phone.

**Pull request merged**